src/poetry/console/application.py

```diff
@@ -277,40 +274,61 @@ def _configure_custom_application_options(self, io: IO) -> None:
                 is_directory=True,
             )
 
-    def _configure_io(self, io: IO) -> None:
-        # We need to check if the command being run
-        # is the "run" command.
-        definition = self.definition
-        with suppress(CleoError):
-            io.input.bind(definition)
-
-        name = io.input.first_argument
-        if name == "run":
-            from poetry.console.io.inputs.run_argv_input import RunArgvInput
-
-            input = cast("ArgvInput", io.input)
-            run_input = RunArgvInput([self._name or "", *input._tokens])
-            # For the run command reset the definition
-            # with only the set options (i.e. the options given before the command)
-            for option_name, value in input.options.items():
-                if value:
-                    option = definition.option(option_name)
-                    run_input.add_parameter_option("--" + option.name)
-                    if option.shortcut:
-                        shortcuts = re.split(r"\|-?", option.shortcut.lstrip("-"))
-                        shortcuts = [s for s in shortcuts if s]
-                        for shortcut in shortcuts:
-                            run_input.add_parameter_option("-" + shortcut.lstrip("-"))
-
-            with suppress(CleoError):
-                run_input.bind(definition)
-
-            for option_name, value in input.options.items():
-                if value:
-                    run_input.set_option(option_name, value)
+    def _configure_run_command(self, io: IO) -> None:
+        """
+        Configures the input for the "run" command to properly handle cases where the user
+        executes commands such as "poetry run -- <subcommand>". This involves reorganizing
+        input tokens to ensure correct parsing and execution of the run command.
+        """
+        command_name = io.input.first_argument
+        if command_name == "run":
+            original_input = cast(ArgvInput, io.input)
+            tokens: list[str] = original_input._tokens
+
+            if "--" in tokens:
+                # this means the user has done the right thing and used "poetry run -- echo hello"
+                # in this case there is not much we need to do, we can skip the rest
+                return
+
+            command_index = tokens.index(command_name)
+
+            # fetch tokens after the "run" command
+            tokens_without_command = tokens[command_index + 1 :]
```

Comment:
I think obtaining the command index like this is incorrect. It fails in degenerate cases like the following directory layout, where we have a separate token called `run` which is not the command:

```
workdir/
  run/
    pyproject.toml
```

```console
$ poetry --project run run echo hi
Command not found: run

$ poetry --project=run run echo hi
Specified path '/tmp/workdir/r' is not a valid directory.
```

(Tested with commit 7dcc5d54)

Reply:
Oops, lets see if we can fix it.

Reply:
Okay have a fix for it, let me know if you find other issues.